File: exports/serve.js
const https = require('https');
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

const dir = __dirname;
const cert = path.join(dir, 'cert.pem');
const key = path.join(dir, 'key.pem');

if (!fs.existsSync(cert) || !fs.existsSync(key)) {
  console.log('Generating self-signed certificate...');
  execSync(`openssl req -x509 -newkey rsa:2048 -keyout "${key}" -out "${cert}" -days 365 -nodes -subj "/CN=localhost"`);
}

const MIME = {
  '.html': 'text/html', '.js': 'application/javascript',
  '.wasm': 'application/wasm', '.pck': 'application/octet-stream',
  '.png': 'image/png', '.svg': 'image/svg+xml',
};

const port = process.argv[2] || 8080;

https.createServer({ cert: fs.readFileSync(cert), key: fs.readFileSync(key) }, (req, res) => {
  let url = req.url === '/' ? '/index.html' : req.url;
  const filePath = path.join(dir, url);
  if (!filePath.startsWith(dir)) { res.writeHead(403); res.end(); return; }

  fs.readFile(filePath, (err, data) => {
    if (err) { res.writeHead(404); res.end('Not found'); return; }
    const ext = path.extname(filePath);
    res.writeHead(200, {
      'Content-Type': MIME[ext] || 'application/octet-stream',
      'Cross-Origin-Opener-Policy': 'same-origin',
      'Cross-Origin-Embedder-Policy': 'require-corp',
      'Cache-Control': 'no-cache',
    });
    res.end(data);
  });
}).listen(port, '0.0.0.0', () => {
  console.log(`Serving on https://0.0.0.0:${port}`);
});

File: exports/serve.py
#!/usr/bin/env python3
"""HTTPS server for Godot web exports with required COOP/COEP headers."""
from http.server import HTTPServer, SimpleHTTPRequestHandler
import ssl, subprocess, os, sys, threading

class GodotHandler(SimpleHTTPRequestHandler):
    def end_headers(self):
        self.send_header("Cross-Origin-Opener-Policy", "same-origin")
        self.send_header("Cross-Origin-Embedder-Policy", "require-corp")
        self.send_header("Cache-Control", "no-cache")
        super().end_headers()

    def log_message(self, format, *args):
        sys.stderr.write("%s - - [%s] %s\n" %
                         (self.address_string(), self.log_date_time_string(),
                          format % args))

def ensure_cert():
    base = os.path.dirname(os.path.abspath(__file__))
    cert = os.path.join(base, "cert.pem")
    key = os.path.join(base, "key.pem")
    if not os.path.exists(cert) or not os.path.exists(key):
        print("Generating self-signed certificate...")
        subprocess.run([
            "openssl", "req", "-x509", "-newkey", "rsa:2048",
            "-keyout", key, "-out", cert,
            "-days", "365", "-nodes",
            "-subj", "/CN=localhost"
        ], check=True)
    return cert, key

os.chdir(os.path.dirname(os.path.abspath(__file__)))
cert, key = ensure_cert()
port = int(sys.argv[1]) if len(sys.argv) > 1 else 8080

server = HTTPServer(("0.0.0.0", port), GodotHandler)
ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
ctx.load_cert_chain(cert, key)
server.socket = ctx.wrap_socket(server.socket, server_side=True)

print(f"Serving on https://0.0.0.0:{port}")
server.serve_forever()
